inject-generator/src/main/java/io/avaje/inject/generator/TypeExtendsReader.java

```diff
@@ -148,59 +155,73 @@ void process(boolean forBean) {
   }
 
   private String initProvidesAspect() {
-    for (String providesType : providesTypes) {
+    for (final String providesType : providesTypes) {
       if (Util.isAspectProvider(providesType)) {
         return Util.extractAspectType(providesType);
       }
     }
     return "";
   }
 
-  private void addSuperType(TypeElement element) {
+  private void addSuperType(TypeElement element, TypeMirror mirror) {
     readInterfaces(element);
-    final String fullName = element.getQualifiedName().toString();
-    if (!fullName.equals(JAVA_LANG_OBJECT) && !fullName.equals(JAVA_LANG_RECORD)) {
+    final String fullName = mirror.toString();
+    if (!JAVA_LANG_OBJECT.equals(fullName) && !JAVA_LANG_RECORD.equals(fullName)) {
       final String type = Util.unwrapProvider(fullName);
       if (isPublic(element)) {
-        extendsTypes.add(type);
+    	 final var genericType= GenericType.parse(type);
+        // check if any unknown generic types are in the parameters (T,T2, etc.)
+        final var knownType =
+        		genericType.params().stream()
+                .flatMap(g -> Stream.concat(Stream.of(g), g.params().stream()))
+                .noneMatch(g -> element(g.mainType()) == null);
+
+        extendsTypes.add(knownType ? type : genericType.topType());
+
         extendsInjection.read(element);
       }
-      addSuperType(superOf(element));
-    }
-  }
 
-  private TypeElement superOf(TypeElement element) {
-    return (TypeElement) asElement(element.getSuperclass());
+      final var superMirror = element.getSuperclass();
+      final TypeElement superElement = asElement(superMirror);
+      addSuperType(superElement, superMirror);
+    }
   }
 
   private void readInterfaces(TypeElement type) {
-    for (TypeMirror anInterface : type.getInterfaces()) {
+    for (final TypeMirror anInterface : type.getInterfaces()) {
       if (isPublic(asElement(anInterface))) {
         readInterfacesOf(anInterface);
       }
     }
   }
 
   private void readInterfacesOf(TypeMirror anInterface) {
-    String rawType = Util.unwrapProvider(anInterface.toString());
+    final String rawType = Util.unwrapProvider(anInterface.toString());
     if (JAVA_LANG_OBJECT.equals(rawType)) {
     } else if (rawType.indexOf('.') == -1) {
       logWarn("skip when no package on interface " + rawType);
     } else if (Constants.AUTO_CLOSEABLE.equals(rawType) || Constants.IO_CLOSEABLE.equals(rawType)) {
       closeable = true;
     } else {
+      final var genericType = GenericType.parse(rawType);
       if (qualifierName == null) {
-        String mainType = GenericType.removeParameter(rawType);
+        final String mainType = genericType.topType();
```

Comment:
What is the thinking behind this change here?

Reply:
reusing the same generic type instance

Reply:
No actual difference between the 2 here - no effective change